blog/02-02-2023-on-chain-randomness-with-celo-using-subgraphs/on-chain-randomness-with-celo-using-subgraphs.md

```diff
@@ -0,0 +1,1859 @@
+
+
+## Introduction
+
+In this tutorial, we will explore how subgraphs and randomness work in the Celo blockchain. This tutorial will cover the use of subgraph technology to monitor blockchain activity and how to integrate random elements into your blockchain application. By the end of this lesson, you'll have a foundational understanding of how subgraphs and randomness can enhance your blockchain application and make it more engaging.
+
+
+## Prerequisites
+
+This tutorial assumes prior knowledge of Hardhat and the Solidity programming language. Its content will focus on practical examples and implementation details rather than introductory explanations. If you are not familiar with these tools, we recommend that you first learn about them before starting this tutorial to fully benefit from its content.
+
+
+## Requirements
+
+-   [Node.js](https://nodejs.org/en/download/) v14.17.6 LTS or higher
+-   [Hardhat](https://hardhat.org/getting-started/#overview)
+-   [OpenZeppelin](https://docs.openzeppelin.com/contracts/4.x/)
+
+
+## Step 1: Understanding Celo Randomness & Subgraphs
+
+### Celo Randomness
+
+Celo's Randomness protocol generates decentralized and transparent random values using a trustless approach. It aggregates random numbers from multiple nodes using a tamper-proof Verifiable Random Function (VRF) for high security, unpredictability, and fairness. Other security and fairness features include a threshold signature scheme and a commit-reveal scheme. The protocol is ideal for various applications, including lotteries, games, and other applications requiring trustworthy randomness. if you want to learn more about Celo Randomness, you can read the [documentation](https://docs.celo.org/protocol/randomness) here.
+
+
+### Subgraphs
+
+The Subgraph protocol is a tool for developers to query data from decentralized networks like Ethereum and IPFS. It enables the creation and publication of public APIs called subgraphs, which index data from decentralized networks for easy accessibility by other developers.Using the Subgraph protocol, developers can create and deploy subgraphs that index data from smart contracts on Ethereum or files on IPFS. These subgraphs can be shared publicly for others to query using GraphQL, a powerful API query language. You can learn more about Subgraphs [documentation](https://thegraph.com/docs/en/cookbook/quick-start/) here.
+
+
+## Step 2: Project Setup
+
+Open a terminal and run the following command to create a new project named LotteryClub and initialize it as a Node.js project.
+
+```bash
+mkdir LotteryClub
+
+cd LotteryClub
+
+npm init -y
+```
+
+Next, install the Hardhat development environment and the OpenZeppelin smart contract libraries by running the following commands:
+
+```bash
+npm install --save-dev hardhat
+
+npm install dotenv @openzeppelin/contracts --save
+```
+
+After the installation is complete, run the following command to initialize the Hardhat project:
+
+```bash
```

Comment:
This doesnt run unless you ask the reader to install "nomicfoundation/hardhat-toolbox" first

Reply:
so maybe add a code snippet before this one with a comment telling the reader to install "'@nomicfoundation/hardhat-toolbox'" first

Reply:
I installed Hardhat by following the instructions in the documentation.

Reply:
We can install nomicfoundation/hardhat-toolbox when we run npx hardhat